cfme/common/provider.py

```diff
@@ -1090,6 +1090,102 @@ def cleanup_vm(vm_name, provider):
         logger.warning('Failed to clean up VM %s on provider %s', vm_name, provider.key)
 
 
+def _is_vm_suspended(provider, vm_name):
+    return provider.mgmt.can_suspend and provider.mgmt.is_vm_suspended(vm_name)
+
+
+def _is_vm_paused(provider, vm_name):
+    return provider.mgmt.can_pause and provider.mgmt.is_vm_paused(vm_name)    
+
+
+def _is_vm_suspended_or_paused(provider, vm_name):
+    return (provider.mgmt.can_suspend and provider.mgmt.is_vm_suspended(vm_name) or
+            provider.mgmt.can_pause and provider.mgmt.is_vm_paused(vm_name))
+
+
+def ensure_vm_state(provider, vm_name, state):
+    """
+    Ensures that the VM/instance is in proper state on provider using provider API.
+
+    State can be one of: "running"/"started", "stopped", "suspended", or "paused"
+
+    Each desired state requires various checks/steps to ensure we have "achieved" that state.
+
+    The logic implied while waiting is:
+    1. Check if VM is in_desired_state, if so, we're done
+    2. If not, check if it is in_state_requiring_prep
+         (for example, you can't stop a suspended VM, so we need to prep the VM by starting it)
+    3. Move the VM to the correct 'prep state' (calling method do_prep)
+    4. Check if the VM is in a state that allows moving to the desired state
+         (using in_actionable_state check)
+    5. Perform the action to put the VM into that state (calling method do_action)
+
+    The methods are defined differently for each desired state.
+
+    For some states, step 2 and 3 do not apply, see for example when desired state is 'running',
+    in those cases just make sure that 'in_state_requiring_prep' always returns false.
+
+    Args:
+        provider: Provider class object
+        vm_name: Name of the VM/instance
+        state: str, one of "running"/"started", "stopped", "suspended", "paused"
+    """
+    vm = vm_name
+
+    if state in ["running", "started"]:
```

Comment:
this begs for a future refactor in wrapanapi

Reply:
I agree